### PR TITLE
Add span-based TimeSpan {Try}Parse{Exact} and TryFormat methods

### PR DESCRIPTION
**Pull request by stephentoub**
This PR:
1. Changes the existing TimeSpan string-based {Try}Parse{Exact} to be span-based.
2. Adds new span-based {Try}Parse{Exact} implementations on top of those same implementations, and also a TryFormat method on top of the same implementation used by ToString.

In doing (1), we not only share code for the string-based and span-based methods, but it also addresses existing allocations in TimeSpan.{Try}Parse.  Today, a call like `TimeSpan.Parse("12:34:56.789")` will allocate ~310 bytes for 3 strings, a string[], and a TimeSpanToken[].  With these changes, TimeSpan.Parse is now allocation-free, as spans and slices are used instead of strings and substrings.  However, the switch from string to span also resulted in some measurable throughput regressions; the second commit represents fixes for some low-hanging fruit that brings the implementations back up to at least throughput parity, and in some cases, makes them faster.

Contributes to https://github.com/dotnet/corefx/issues/22375
cc: @AlexGhiondea, @joperezr, @jkotas, @tarekgh

### CODE REVIEW

File: src/mscorlib/src/System/Globalization/TimeSpanParse.cs [hunk 1]
```diff
@@ -419,6 +420,9 @@ internal Boolean FullHMSFMatch(TimeSpanFormat.FormatLiterals pattern)
             private const int MaxLiteralTokens = 6;
             private const int MaxNumericTokens = 5;
 
+            internal TimeSpanToken numbers0, numbers1, numbers2, numbers3, numbers4; // MaxNumbericTokens = 5
+            internal ReadOnlySpan<char> literals0, literals1, literals2, literals3, literals4, literals5; // MaxLiteralTokens=6
```
Comment (stephentoub) [inline src/mscorlib/src/System/Globalization/TimeSpanParse.cs]: @jaredpar, per our conversation, this is an example of where fixed buffer support for non-primitive types would be really handy.  You can see here:
https://github.com/dotnet/coreclr/pull/13619/commits/4ad7260d872b2356262337b15f854bddd7bbe01d#diff-e581a221cef607379517b80100253877R486
having to need to implement our own lame "indexer", and then all of the changes starting around:
https://github.com/dotnet/coreclr/pull/13619/commits/4ad7260d872b2356262337b15f854bddd7bbe01d#diff-e581a221cef607379517b80100253877R927
that are necessary to target the individual fields, making for a less-than-ideal porting experience from strings to spans.
cc: @MadsTorgersen

File: src/mscorlib/src/System/Globalization/TimeSpanFormat.cs [hunk 2]
```diff
@@ -36,6 +36,29 @@ internal enum Pattern
         // 
         internal static String Format(TimeSpan value, String format, IFormatProvider formatProvider)
```
Comment (jkotas) [inline src/mscorlib/src/System/Globalization/TimeSpanFormat.cs]: It maybe useful to move TimeSpanFormat.cs and TimeSpanParse.cs to shared first so this change ports to CoreRT for free. The CoreCLR and CoreRT copies are not that far off.

Comment (stephentoub) [reply src/mscorlib/src/System/Globalization/TimeSpanFormat.cs]: Ugh, somehow I managed to convince myself that all of the files were already shared :smile:

Comment (stephentoub) [reply src/mscorlib/src/System/Globalization/TimeSpanFormat.cs]: I've added some commits to move the files to shared.

File: src/mscorlib/src/System/Globalization/TimeSpanParse.cs [hunk 3]
```diff
@@ -157,71 +142,93 @@ public bool IsInvalidNumber(int maxValue, int maxPrecision)
...
+                if (pos >= m_value.Length)
                 {
-                    tok.ttt = TTT.End;
-                    return tok;
+                    return new TimeSpanToken(TTT.End);
                 }
 
-                if (ch >= '0' && ch <= '9')
+                // Now retrieve that character. If it's a digit, we're processing a number.
+                int num = m_value[pos] - '0';
```
Comment (tarekgh) [inline src/mscorlib/src/System/Globalization/TimeSpanParse.cs]: nit: maybe we can simplify the code for the leading zeros here by just doing 

while (m_pos <= m_value.Length && m_value[m_pos] == '0')
    m_pos++;

if (m_pos >= m_value.Length || (uint)(m_value[m_pos] - '0') > 9)
   return new TimeSpanToken(TTT.Num, 0, m_pos - pos, default(ReadOnlySpan<char>));

The current code looks ok too.

**Comment by tarekgh** [comment]
LGTM.

File: src/mscorlib/shared/System/Globalization/TimeSpanParse.cs [hunk 4]
```diff
@@ -1592,160 +1509,158 @@ internal bool TryParse(ReadOnlySpan<char> input, ref TimeSpanResult result)
...
+
                 if (negative)
                 {
                     time = -time;
                     // Allow -0 as well
                     if (time > 0)
                     {
-                        result.SetFailure(ParseFailureKind.Overflow, "Overflow_TimeSpanElementTooLarge");
-                        return false;
+                        return result.SetFailure(ParseFailureKind.Overflow, SR.Overflow_TimeSpanElementTooLarge);
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/TimeSpanParse.cs]: Using `SR.Overflow_TimeSpanElementTooLarge` for the failure case will make the TryXXX methods a lot slower when they return false. This should stay on the CoreCLR plan.

Comment (jkotas) [reply src/mscorlib/shared/System/Globalization/TimeSpanParse.cs]: `SR.Overflow_TimeSpanElementTooLarge` will fetch the resource string. It is not a cheap operation.

Comment (stephentoub) [reply src/mscorlib/shared/System/Globalization/TimeSpanParse.cs]: Ah, makes sense.  I'll switch this back from the corert approach to the coreclr one.

Comment (stephentoub) [reply src/mscorlib/shared/System/Globalization/TimeSpanParse.cs]: I've fixed this.  Thanks for the catch.

**Comment by stephentoub** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test please ("Error fetching remote repo 'origin'"... FYI, @dotnet/dnceng)

**Pull request merged**